Spawn candelabra

eh-jogos opened the pull request:
Should solve the candelabra part of https://github.com/Mooses2k/SecretHistories/issues/21#issuecomment-1537558915

The first two commits are a kind of unrelated to this PR and maybe should have gone in another one, they just fix a bug with map printing and add script_templates with some "pretty guidelines" as discussed in Discord.

## RoomData
A resource that holds common and useful data for rooms. Mainly it's original `rect2`, it's `cell_indexes`, and has properties for `type` and `polygons`, which is what the `rooms` property was being saved for.

The `rooms` property was changed to store a Dictionary with room_types as keys and an Array of RoomData as value, instead of an Array of polygons.

For now I'm only categorizing rooms into "starting_room" and "generic".

With this addition, I think it should be possible to remove the `STARTING_ROOM` enum I added a few PRs ago, as now it's possible to get the starting room data at any type, and it should actually be easier to do it this way.

## GeneratorGroup
Also as discussed on Discord, I added a `GeneratorGroup` class that extends `GeneratorStep`, and will loop through it's own child steps or child groups before proceeding to the next step.

## Candelabra Spawn Step
The next commits are various steps in getting candelabra spawning to work. As discussed in Discord, the `GenerateCandelabra` step holds all the logic for when and where they should spawn, and I've renamed `LootSpawnList` to a more generic `ObjectSpawnList` and made `LootData` into its own class called `SpawnData.

`GenerateCandelabra` will add fully configured `SpawnData` to `WorldData`, and after world generation is finished we loop through the list of objects to spawn and use `SpawnData` to spawn them. 

For now some logic is repeated on ItemSpawner and SpawnData, but I think SpawnData is the better place for them as it can be more easily reused anywhere. And the steps for spawning cultist and initial loot I've done in previous PRs should be refactored to follow this organization. The logic is mostly ready so it should be a case of moving logic to better places such as GenerationSteps (probably with some few adpatations), and deleting duplicated code from ItemSpawner anc CharacterSpawner.

## Fixing RNG
While cultist and loot spawn is not refactored, I just fixed where they were using either global random functions or RandomNumberGenerators that didn't use the shared seed for the world.

crazyStewie commented on scenes/worlds/world_gen/generation_steps/generation_group.gd:
```diff
@@ -0,0 +1,49 @@
+
+### Public Methods --------------------------------------------------------------------------------
+
+func execute_step(data : WorldData, gen_data : Dictionary, generation_seed : int):
```
Please avoid overriding function not marked for that unless strictly necessary, to avoid having to manually keep behavior in sync when changing it on the base class (for example, if GenerationStep starts measuring its own time and emitting the step finished signals itself, we would have to also update this here), and to avoid surprising behaviors when calling this function without realizing it might have been overriden. In this case, since it seems to do exactly the same as the base class, just removing the override definition should be enough.

eh-jogos replied:
This is here because GenerationGroup didn't extend GenerationStep at the beggining, but later when I changed it forgot to erase it. Will erase it.

I'm not sure GenerationGroup should inherit GenerationStep, but due to some limitations in GDScript compared to GDscript 2.0 it was easier this way.

crazyStewie replied:
Yeah, inheriting GenerationStep is probably the right choice in this case, makes it more modular. since for the caller, there won't be any difference, and we can have any depth of generation steps without issue.

crazyStewie commented on resources/world/world_data.gd:
```diff
@@ -402,6 +442,13 @@ func get_cell_type(cell_index : int) -> int:
 
 func set_cell_type(cell_index : int, value : int):
 	if cell_index >= 0:
+		if cell_type[cell_index] == CellType.STARTING_ROOM and value == CellType.ROOM:
```
This introduces extremely surprising behavior, and should be avoided. Trying to set a cell to a specific value should always set it to that value, unless it's impossible for very strong reasons, in which case, the value should stay the same, but the lack of change should be communicated to the caller somehow (If a situation like that ever appears, this function should be modified to return some error value, or at least a boolean to indicate success)

Since you need to know this information for other things to work, here are a few alternatives:
- Use cell meta, and store the information that the cell is part of the starting room in there (not that good, will need a lot of cells with this data, and modifying the starting room requires modifying all the meta data)

- Make a new function that checks if a cell is within a given room type, or returns the room type of a given cell. This would preferably be done by checking the cell against the rooms Rect2s or Polygons (polygons not implemented yet), instead of checking if the cell is in the cell array. This way, you could check if the cell is in the starting cell when you need this information.

- Make a new dense array (something like a PoolIntArray) that stores, for every cell, the room type they belong to. this would be similar to the cell_type array, but allow for room specific information, where cell_type would be a more generic thing.

eh-jogos replied:
I don't think this is needed anymore since we have the RoomData, but I have to refactor STARTING_ROOM to stop existing and use `rooms["starting_room"]`, but that's outside of the scope of this PR because I'd have to refactor the previous two PRs. Already talked about it with @Mooses2k 

before doing this I checked and this was the only override I wanted to block, so pretty safe for now until I work on the starting room refactor for making cultist spawn and initial loot spawn follow GenerationSteps next week.

crazyStewie commented on resources/world/world_data.gd:
```diff
@@ -429,6 +477,38 @@ func get_cells_for(p_type: int) -> Array:
+
+
+# Checks if a cell index is free
+func is_cell_free(cell_index: int) -> bool:
```
This can be done more efficiently by checking if the _objects_to_spawn dictionary has the cell index as a key using Dictionary.has(), and checking if the cell_index is the same as the player spawn cell index.

eh-jogos replied:
As far as I know `not cell_index in occupied_cells` is syntatic sugar for `has` and not walking into the array, and I remember seeing it recommended over `.has` because of some gdscript optimization but couldn't find it again, so I migth be remembering wrong. 

I'll change this as requested, but I though the benefit of this is that we already have two different conditions for checking if cells are free, and there will possibly be more later, as I think characters should have a different property than `_objects_to_spawn` and later on, with objects that spawn more than one cell, this might have to change again. 

So I was thinking it's easier to read this code and understand it, and we can just keep changing `get_occupied_cells` as the rules change. Plus other places might want to get all occupied cells for other reasons.

crazyStewie replied:
Array.has() in an array walks the array, The array has no guarantees of any order, so it can't do anything efficient like a binary search, so it walks the array until it finds a match.
In the godot source, Array.find() calls Vector.find() inside an internal ArrayPrivate (https://github.com/godotengine/godot/blob/3.5/core/array.cpp#L210)
Which calls CowData.find() (https://github.com/godotengine/godot/blob/3.5/core/vector.h#L90)
Which walks the array (https://github.com/godotengine/godot/blob/3.5/core/cowdata.h#L336)

crazyStewie commented on resources/world/world_data.gd:
```diff
@@ -329,6 +340,35 @@ func is_room_cell(p_index: int) -> bool:
+	if rooms.has(p_type):
+		value = rooms[p_type]
+	else:
+		push_error("Type of room not found: %s. Available types: %s"%[p_type, rooms.keys()])
```
Requesting an array of rooms of a type that isn't present in the map shouldn't be treated as an error (at least by this function). Not having rooms of a certain type is perfectly valid, and this should just return an empty array in that case. If the caller of this function expects the room type to exist, it can then treat it as an error if the array has a size of 0. Also, the Dictionary.get() function allows setting a default value to return if the key isn't present in the dictionary, which would simplify this function to `return rooms.get(p_type, Array())`

crazyStewie commented on scenes/worlds/world_gen/generation_steps/populate_room/generate_candelabra.gd:
```diff
@@ -0,0 +1,115 @@
+		var neighbour_index := world_data.get_cell_index_from_int_position(
+				neighbour_position.x, neighbour_position.y
+		)
+		if world_data.get_cell_type(neighbour_index) == world_data.CellType.DOOR:
```
A celltype of Door isn't a strong guarantee if there is a door or not in that direction, and it makes no guarantees as to whether the door is pointing in the direction of this cell, and not some completely unrelated place. For cases like this, using the wall indices to look for a EdgeType of either Wall or Empty would be preferred (if all edges around this cell are walls or empties, there are no doors present)
The Edge type is also what is used both when generating the navmesh and actually spawning in doors.

Mooses2k reviewed (changes requested):
Minor stuff from me, only mentioned since Stewie's going to request changes..
Min size of considered room should be 4 not 6.
Let's do 85% lit, 10% unlit, 5% missing.

eh-jogos commented:
@Mooses2k @crazyStewie, just rebased this PR and updated it with three commits to address the reviews. In each commit description there's a link to the review comment that it relates to
